File: todo.hs
{-- An simple todolist manager: 
 - Planned features : replace , prioritise
 --}

module Main( main ) where

import System.Environment
import System.Directory
import System.Console.GetOpt
import System.IO
import Data.Maybe( fromMaybe )
import Data.List
import System.Exit

fileName="/home/jelle/todo"

main = do
  args <- getArgs
  let ( actions, nonOpts, msgs ) = getOpt RequireOrder options args
  opts <- foldl (>>=) (return defaultOptions) actions
  let Options { optInput = input,
                optOutput = output } = opts
  input >>= output

data Options = Options  {
    optInput  :: IO String,
    optOutput :: String -> IO ()
  }

defaultOptions :: Options
defaultOptions = Options {
    optInput  = getContents,
    optOutput = putStr
  }

--cli options 
options :: [OptDescr (Options -> IO Options)]
options = [
    Option ['h'] ["help"]    (NoArg showHelp) "show help",
    Option ['v'] ["view"]     (NoArg view ) "view todoList",
    Option ['a'] ["add"]     (ReqArg add "todo" ) "add item to todolist",
    Option ['r'] ["remove"]  (ReqArg remove "lineNumber" ) "remove item from list",
    Option ['s'] ["search"]  (ReqArg search "keyword") "search you todo items",
    Option ['V'] ["version"] (NoArg showVersion)         "show version number"
  	  ]

--show the version of the porgram
showVersion _ = do
  putStrLn "TodoListEditor version 0.1"
  exitWith ExitSuccess

--show usage
showHelp _ = do
  putStrLn "Usage: [-vio] Args"
  exitWith ExitSuccess

--View the todo list
view _ =  do
    contents <- readFile fileName
    let todoTasks = lines contents
        numberedTasks = zipWith (\n line -> show n ++ " - " ++ line) [0..] todoTasks
    putStr $ unlines numberedTasks 
    exitWith ExitSuccess

--Add an todo item
add args opt = do 
	appendFile fileName (args ++ "\n")
	exitWith ExitSuccess	

--Remove an todo item
remove args opt = do
	handle <- openFile fileName ReadMode
	(tempName, tempHandle) <- openTempFile "." "temp"
	contents <- hGetContents handle
	let number = read args :: Int
            todoTasks = lines contents
            newTodoItems = delete (todoTasks !! number) todoTasks
	hPutStr tempHandle ( unlines newTodoItems )
	hClose handle
	hClose tempHandle
	removeFile fileName
	renameFile tempName fileName
	exitWith ExitSuccess	

--Search an todo item
search args opt = do
	contents <- readFile fileName
	let todoTasks = filter (args `isInfixOf`) (lines contents)
	putStr ( unlines todoTasks )
	exitWith ExitSuccess	

File: todo.py
#!/usr/bin/python
import os, sys, string

# the todolist
filename = "/home/jelle/Projects/progsel/todo.txt"

#status

def help_message():
	print '''todo.py -- an simple CLI todolist editor

	Usage:
		-s, 	--search 		search the todofile with max 3 keywords
		-v, 	--view 			view the todofile
		-a, 	--add 			add an item to the todofile
		-r, 	--remove 		remove an line from the todofile'''
	exit(0)

# search
def search(keyword):
	keywords = len(keyword)
	try:
		f = open(filename,"r")
	except IOError:
		print 'cannot open', filename
	else:
		if keywords == 1:
			for line in f:
				if line.rfind(keyword[0]) != -1:
					print line.replace('\n',' ',1)
		elif keywords == 2:
			for line in f:
				if line.rfind(keyword[0]) != -1  and line.rfind(keyword[1]) != -1:
					print line.replace('\n',' ',1)
		elif keywords == 3:
			for line in f:
				if line.rfind(keyword[0]) != -1  and line.rfind(keyword[1]) != -1 and keyword[2] != -1:
					print line.replace('\n',' ',1)

	f.close()


# add numbers to output
# view the todolist
def view():
	try:
		f = open(filename,"r")
	except IOError:
		print 'cannot open', filename
	else:
		x = 1
		for line in f:
			print str(x) + '. ' + line.replace('\n',' ',1)
			x = x + 1
		f.close()



# add an todo item
# argument string
def add(text):
	newline = ''
	for item in text:
		newline += item + ' ' 
	newline += '\n'
	
	try:
		f = open(filename,"a")
	except IOError:
		print 'cannot open', filename
	else:
	#	lines = 1
	#	for line in open(filename,"r"): 
	#		lines = lines + 1
#
#		newline = str(lines) + '. ' + newline 
		f.write(newline)
		f.close()
	
# remove an linenumber
def remove(line):
	line = line -1
	try:
		f = open(filename,"r")
	except IOError:
		print 'cannot open', filename
	else:
		# store the data
		data = f.readlines()
		f.close()

		# modify data
		del data[line]

		# write modified data
		f = open(filename,"w")
		f.writelines(data)
		f.close()



# Parsing of arguments
try:
	if len(sys.argv) == 1:
		help_message()

	arg = sys.argv[1]
	oper = sys.argv[2:]

	if arg == '-s':
		if len(oper) == 0:
			print "please insert an argument please"
		else:
				search(oper)
	elif arg == '-v':
		view()
	elif arg == '-a':
		add(oper)
	elif arg == '-r':
		if len(oper) == 1:
			remove(int(oper))
		else:
			print "arguments needs to be one number."

	else:	
		help_message()
except:
	pass
			




